lib/valueflow.cpp

```diff
@@ -7008,6 +7008,7 @@ static void valueFlowDynamicBufferSize(const TokenList& tokenlist, const SymbolD
         case Library::AllocFunc::BufferSize::strdup:
             if (arg1 && arg1->hasKnownValue()) {
                 const ValueFlow::Value& value = arg1->values().back();
+                assert(value.isKnown());
```

Comment:
@pfultz2 @firewave I don't feel I fully remember the details. do we want to enforce that the known value is always last? if we see this code maybe we should start making it more flexible instead?

Reply:
The code contradicts itself. It checks if there is a known value from the front but then always uses the last value which might *not* be known. The assert is to detect such a case.

Based on the check it should also be calling `getKnownValue()` but I do not want to change that blindly without a case hitting it.

Reply:
If I comment out these lines :
```
                //const ValueFlow::Value& value = arg1->values().back();
                //if (value.isTokValue() && value.tokvalue->tokType() == Token::eString)
                //    sizeValue = Token::getStrLength(value.tokvalue) + 1; // Add one for the null terminator
```
Then this test fails: `TestValueFlow::valueFlowDynamicBufferSize`
So this code is hit by a test case.

Reply:
This is about some input hitting the *wrong* assumption and not the code in general.

Reply:
Done. All tests still pass.